docs/firmware-architecture-and-design.md

```diff
@@ -0,0 +1,119 @@
+# Architecture and Design Rationales
+
+# Table of Contents
+* [General Principles](#general-principles)
+* [App/IO Split](#appio-split) 
+* [Design Patterns](#design-patterns)
+  * [Pseudo-Class](#pseudo-class)
+
+# General Principles
+* Avoid memory allocation wherever possible. This means that we usually try to limit our use of pseudo-classes to fairly high level abstractions.
+* **REALLY** avoid allocating memory anywhere you don't have a guaranteed upper-bound on the number of times that the allocation will be executed. This is so we can have more guarantees on the upper bound of memory usage, and violating this can potentially cause *very* serious bugs.
+* Never use `static` in anything in the `app` layer. As the `app` layer is intended to be portable to our simulator, this means that any `static` variables will be re-used between multiple robots in the simulator, almost guaranteeing undesired behavior.
+
+# Design Patterns
+
+## Pseudo-Class
```

Comment:
nit: this example violates the design guide doesn't have `app`, `io` or `shared` in functions

Reply:
Good point, added `app_` prefix